Foundation/NSURLProtocol.swift

```diff
@@ -142,6 +145,92 @@ public protocol URLProtocolClient : NSObjectProtocol {
     func urlProtocol(_ protocol: URLProtocol, didCancel challenge: URLAuthenticationChallenge)
 }
 
+internal class _ProtocolClient : NSObject, URLProtocolClient {
+
+    func urlProtocol(_ protocol: URLProtocol, didReceive response: URLResponse, cacheStoragePolicy policy: URLCache.StoragePolicy) {
+        `protocol`.task?.setResponse = response
+    }
+
+    func urlProtocolDidFinishLoading(_ protocol: URLProtocol) {
+        guard let session = `protocol`.task?.session as? URLSession else { fatalError() }
+        switch session.behaviour(for: `protocol`.task!) {
+        case .taskDelegate(let delegate):
+            guard let s = session as? URLSession else { fatalError() }
+            s.delegateQueue.addOperation {
+                delegate.urlSession(s, task: `protocol`.task!, didCompleteWithError: nil)
+                `protocol`.task?.state = .completed
+            }
+        case .noDelegate:
+            `protocol`.task?.state = .completed
+        case .dataCompletionHandler(let completion):
+             let data = Data()
+            `protocol`.client?.urlProtocol(`protocol`, didLoad: data)
+            return
+        case .downloadCompletionHandler(let completion):
+            guard let s = session as? URLSession else { fatalError() }
+            s.delegateQueue.addOperation {
+                completion(`protocol`.task?.currentRequest?.url, `protocol`.task?.response, nil)
+                `protocol`.task?.state = .completed
+            }
+        }
+    }
+
+    func urlProtocol(_ protocol: URLProtocol, didCancel challenge: URLAuthenticationChallenge) {
+        NSUnimplemented()
+    }
+
+    func urlProtocol(_ protocol: URLProtocol, didReceive challenge: URLAuthenticationChallenge) {
+        NSUnimplemented()
+    }
+
+    func urlProtocol(_ protocol: URLProtocol, didLoad data: Data) {
+        guard let session = `protocol`.task?.session as? URLSession else { fatalError() }
+        switch session.behaviour(for: `protocol`.task!) {
+        case .dataCompletionHandler(let completion):
+            guard let s = `protocol`.task?.session as? URLSession else { fatalError() }
+            s.delegateQueue.addOperation {
+                completion(data, `protocol`.task?.response, nil)
+                `protocol`.task?.state = .completed
+            }
+        default: return
+        }
+    }
+
+    func urlProtocol(_ protocol: URLProtocol, didFailWithError error: Error) {
```

Comment:
nit: Will it be possible to factor out some code this function shares with `urlProtocol(_ protocol: URLProtocol, didFailWithError error: Error)`?